left_sidebar: Prioritize unmuted topics in non-zoomed topic list.

Ddharmani3 opened the pull request:
This PR fixes part 4 of the Left sidebar section of https://github.com/zulip/zulip/issues/24243:
- [x] Show unmuted topics within a stream with priority over all muted topics, i.e., you'll see them at the top regardless of age in the default view before you click "more topics", even if muted topics have unreads. This will match what happens if you have some muted topics in a stream where topics are by default "not muted".


Created a new function `choose_topics` that loops through the topics and push filtered topics using `should_show_topic` function to `items` array if not zoomed else just push all topics directly to array.

If stream is muted and not zoomed call the `choose_topics` function twice, first with passing `unmuted_topics` and second time with passing remaining topics. else, call it only once with `topic_names`.

<!-- Describe your pull request here.-->

Fixes: <!-- Issue link, or clear description.-->
#24243
<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
https://user-images.githubusercontent.com/64723994/231698353-b1737eb4-cbd8-4c04-9d83-2196427ca812.mp4

example case: In muted stream(ビデオゲーム), Topics 1 to 10 are unmuted, topic 1 doesn't have unread hence it only shows 2-10(9 topics), and topics 14,15,16 are muted(inherit) and has unread hence are also shown(3topics) = 12 topics + 1 active topic(FAILED EXPORT). some other inherit topics have unread messages but are not shown as max_topics_with_unread limit reached.

- **!Zoomed:** 

![image](https://user-images.githubusercontent.com/64723994/233417581-b9e99656-c615-4275-ab24-1d296e99ef00.png)

- **Zoomed:**

![image](https://user-images.githubusercontent.com/64723994/233417882-55d909d4-3540-4b5b-b3eb-a183362ad437.png)


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

timabbott commented on web/tests/topic_list_data.test.js:
```diff
@@ -250,6 +255,12 @@ test("get_list_info unreads", ({override}) => {
+    override(user_topics, "is_topic_unmuted", (stream_id, topic_name) => {
+        assert.equal(stream_id, general.stream_id);
+        return topic_name === "topic 5";
+    });
```
This is fine for now but we should as a follow-up rewrite this test to just use the actual `user_topics` module and configure a set of data; this kind of hacky mock replacing a data module with a rule like this is generally to be avoided.

timabbott commented on web/src/topic_list_data.js:
```diff
@@ -106,12 +113,25 @@ export function get_list_info(stream_id, zoomed, search_term) {
+                unmuted_topics.push(topic_info);
+            } else {
+                normal_topics.push(topic_info);
+            }
```
It looks like this changes the sort order, not just whether or not the topic is displayed; is that accurate? 

I think that behavior is wrong is the stream isn't muted, so you need a check for whether the stream is muted in this block somehow -- an unmuted topic in a stream that is not muted should be a noop, not reorder to the top of the left sidebar for the stream.

Ddharmani3 replied:
> It looks like this changes the sort order, not just whether or not the topic is displayed; is that accurate?

Yes this part of the code only sorts the order but the `should_show_topic()` function decides whether or not the topic is displayed to which I have added the condition to ensure that unmuted topics are displayed by checking is_topic_unmuted that returns true.

> I think that behavior is wrong is the stream isn't muted, so you need a check for whether the stream is muted in this block somehow -- an unmuted topic in a stream that is not muted should be a noop, not reorder to the top of the left sidebar for the stream.

I have updated the PR to ensure that the sorting only occurs when the stream is muted. Although there is currently no option to unmute a topic in an unmuted stream, I have added an if condition to ensure that the sorting is only done in muted streams. This condition ensures that topics are not sorted in an unmuted stream, even if there are some unmuted topics in any case.

timabbott replied:
Yeah the order of operations to get into that state would be first unmuting the topic, and then unmuting the stream.

alya commented:
I don't think this works right in the "more topics" view. We normally don't reorder topics in that view based on whether or not they've been muted, and I think we should stick with that.

Ddharmani3 commented:
> I don't think this works right in the "more topics" view. We normally don't reorder topics in that view based on whether or not they've been muted, and I think we should stick with that.

The behavior remains unchanged. In the "more topics" view, the topics are not rearranged and remain in the original order, as I only sort the order when the topic list is not zoomed (i.e., not in the "more topics" view). 

https://user-images.githubusercontent.com/64723994/231734202-75ce190b-ab13-4cf5-aca1-2fc93ee3f1e3.mp4

In the video above, you can see the initial state of the topics. I then unmuted one topic and muted another. However, when I clicked on "more topics," they remained in the same order as the initial state.

alya commented:
@Ddharmani3 What is the state of addressing @timabbott 's feedback on this one?

Ddharmani3 commented:
> @Ddharmani3 What is the state of addressing @timabbott 's feedback on this one?

@alya I have addressed the feedback from @timabbott to ensure that reordering is only done in a muted stream. This PR is ready for review now.

alya commented:
OK, great, I believe this is working as expected in manual testing. @timabbott do you want to take another look?

timabbott commented on web/src/topic_list_data.js:
```diff
@@ -62,6 +67,10 @@ export function get_list_info(stream_id, zoomed, search_term) {
 
+                if (is_topic_unmuted) {
+                    return true;
+                }
```
This algorithm is wrong. First, unmuted topics should have no special priority in a stream that is not muted, and this code does not do that. Further, if you have 25 unmuted topics in a stream, this will show all of them in the !zoomed case, since this check is before the `idx < max_topics` one. I don't see an easy way to fix that with this single pass iteration through the topics algorithm like what you have here, because the (inherit) topics have no way to know if we'll later find 10 (old!) unmuted topics, and thus the 6th oldest topic should have been skipped for that reason.

I think the correct thing to do is turn this `for` loop into a function (`choose_topics`) that accepts the list of topics as a parameter. For a stream that is not muted, we just call the function with `topics_names`. For a muted stream, we call that function twice -- first with just the unmuted topics, and then with all the other topics, and then concatenate the results.

It'll need some careful preparatory refactoring; the `idx < max_topics` check probably needs to be `idx < max_topics && topics_selected < max_topics`, since otherwise we will overselect in the second pass if there are several unmuted topics older than the 8th oldest topic.

We'll also need to be careful to ensure the `topics_selected` value is shared between the two loop runs; basically any unmuted topics we've selected should take away from how many (inherit) topics in the muted stream we display after clicking into the stream.

Ddharmani3 commented:
@timabbott can you please review whether I have executed the approach you mentioned properly?

timabbott commented on web/src/topic_list_data.js:
```diff
@@ -3,37 +3,26 @@ import * as resolved_topic from "../shared/src/resolved_topic";
-
+    const topics_with_unread_mentions = unread.get_topics_with_unread_mentions(
+        stream_info.stream_id,
+    );
```
Let's pass this value in from the caller -- there's no reason to compute it twice, which is what your algorithm here does.

timabbott commented on web/src/topic_list_data.js:
```diff
@@ -107,17 +96,50 @@ export function get_list_info(stream_id, zoomed, search_term) {
+}
+
+export function get_list_info(stream_id, zoomed, search_term) {
+    const stream_information = {
```
`stream_info` is a questionable name for this -- it's not really information about the stream. I'm not sure what a great name is, though. 

(As a side note, you should almost never have a thing you call `stream_information` in one place and `stream_info` in the other -- it's way way easier to read code if a variable/value doesn't get randomly renamed between different abbreviations across a code path. I think I've pointed this out in your work before -- please pay attention to details like this! Whenever you've written code that you think works, you should do another read of it looking at what you called things and make sure it feels good).

timabbott replied:
I think my proposal is that you make `stream_id` not part of this data structure -- it can be just the first parameter to `choose_topics` -- and `items` could be added to this data structure rather than an independent parameter -- and then this can be `topic_choice_state`.

timabbott commented on web/tests/topic_list_data.test.js:
```diff
@@ -259,11 +268,11 @@ test("get_list_info unreads", ({override}) => {
             "topic 1",
             "topic 2",
             "topic 3",
-            "topic 5",
```
I think the algorithm implementation is probably correct at this point, but what you should do is extend the tests such that the tests would fail with your previous implementation but pass with the new one -- i.e actually exercise the corner cases that the previous version handled incorrectly.

timabbott commented:
> can you please review whether I have executed the approach you mentioned properly?

This is great progress and I think the correct algorithm, but needs some cleanup and verification -- posted a few comments above that should help you get this into shape. If you're having trouble writing the test, we can possibly defer that for a follow-up, but please at least manually do data setup in the development environment (i.e. send messages to 22 topics in a stream and unmuted half of them) and make sure it handles the cases we discussed previously correctly -- I don't see confirmation you've done that.

Ddharmani3 commented:
@timabbott addressed the above feedback in the updated PR. This PR is ready for review now.

> make sure it handles the cases we discussed previously correctly -- I don't see a confirmation you've done that

I had tested it but forgot to mention that :( . Here is the SS for one case: 
Info: Topics 1 to 10 are unmuted, topic 1 doesn't have unread hence it only shows 2-10(9 topics), and topics 14,15,16 are muted(inherit) and has unread hence are also shown(3topics) = 12 topics + 1 active topics(FAILED EXPORT). some other inherit topics have unread messages but are not shown.
![image](https://user-images.githubusercontent.com/64723994/233411681-cbdde419-0ed2-4b11-8f36-472f1a61a24c.png)

>  If you're having trouble writing the test, we can possibly defer that for a follow-up, 

I will create a follow-up PR to add tests as soon as this PR is merged. As I am new to writing frontend tests, there may be some feedback rounds, which could potentially hold up the current PR. Therefore, to ensure that this "release merge" PR is not blocked, I believe it's best to write the tests as a separate follow-up PR.

timabbott commented:
Sounds good, merged, thanks @Ddharmani3! You might fight @showell's videos helpful for learning how to write node tests, since he's the original author of that test system.

(One observation is that it might be nice to try reworking the tests in general to mostly test `choose_topics`, since most of the calls to outside modules are not there, so it could be done with a lot less mocking).
